tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,511 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunks
+
+import (
+	"bufio"
+	"encoding/binary"
+	"io"
+	"io/ioutil"
+	"math"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+// Segment header fields constants.
+const (
+	headChunksFormatV1 = 1
+)
+
+// HeadReadWriter is for writing the Head block chunks to the disk
+// and access chunks via an mmapped file.
+type HeadReadWriter struct {
+	// Writer.
+	dirFile          *os.File
+	curFile          *os.File // Segment file being written to.
+	curFileStartTime time.Time
+	curFileSequence  int
+	wbuf             *bufio.Writer
+	wbufLock         sync.Mutex
+	n                int64 // Bytes written in current segment.
+	buf              [8]byte
+	segmentTime      time.Duration
+
+	// Reader
+	bs    map[int]ByteSlice // TODO: this should be a map of the file name and bytes slice as old slices will be deleted.
+	cs    map[int]io.Closer // Closers for resources behind the byte slices.
+	size  int64             // The total size of bytes in the reader.
+	bsMtx sync.RWMutex      // TODO: verify the usage of this
+	pool  chunkenc.Pool
+}
+
+const (
+	// DefaultHeadChunkSegmentTime is the default chunks segment time range.
+	DefaultHeadChunkSegmentTime = 15 * time.Minute
```

Comment:
:+1: 

I would add a comment why 15 minutes.

Reply:
Thinking more about it, maybe 15m isn't right. Assuming an average scrape interval of 15s, the chunks would be completed after 30m, so with 15m as the default, there would be many empty files. I will make it 40m with the reasoning in the comments.